### PR TITLE
[Manh Duc] iP

### PR DESCRIPTION
**Pull request by hmanhduc2k**
# DuC (That time I create a chatbot with my name for CS2103T IP because Singaporeans keep mispronouncing it then asking me how  :joy::joy::thinking::vomiting_face:)

Yea you heard me right. The most generic yet difficult-to-pronounce chatbot you will have ever used. Why use me then?
Because I am **generic**, it means that there is absolutely no risk of using me. I can be the most _useful~~~_ chat bot you will ever use without even notice. Because I'm doing my job properly and not trying to market you with distrustful stuffs.

Most importantly, please learn how to pronounce my name after using DuC. Concluded with thanks.

So, enough ~~hidden marketing~~ long-winding introduction that sound like a rip-off light novel title already. What can I offer to you?

1. Schedule a to-do task
2. Schedule a deadline task with prescribed date
3. Schedule an event task with occurred date
4. Track task or delete task on completion - of your choice
5. Find task by date or by keywords

Does that sound too generic? I also offer this:
* Easy to use and remember command line
* Difficult to pronounce names that will make you question how to pronounce it 100% of the time
* Some other things because CS2103T requires so

How does DuC sound like to you? Here is a sample list of daily schedule you can follow with DuC
- [x] [T] code
- [ ] [D] code while sleeping (by everyday)
- [ ] [T] code while eating
- [ ] [T] code while taking a sh*t
- [ ] [E] code while crying in the corner of your dorm room because ~~you have no social life and are a miserable victim of crony capitalism and materialism~~ you cannot contain excessive joy in doing programming

How about a glimpse of my development process?
``` java
case "todo":
      return new TaskCommand(description, taskList, Task.Type.TODO);
case "deadline":
      return new TaskCommand(description, taskList, Task.Type.DEADLINE);
case "event":
      return new TaskCommand(description, taskList, Task.Type.EVENT);
```

Basically, there are three types of task you can add to your list:
`<todo>`,`<deadline>`,`<event>`. Sounds very generic is it? I know

Here is a random filler quote because I don't know what else to add:
> Death is like the wind. Always by my side - Yasuo

To conclude, please click [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) if you are interested to find out more about my project. 

Nah I'm just messing with you, do click [here](https://github.com/hmanhduc2k/ip/releases/download/v01/DuC-1.0-all.jar) instead to try out my product.

Peace~~~

### CODE REVIEW

File: src/main/java/Duke/Command/DateOfCommand.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import Duke.Main.TaskList;
+import Duke.Task.Deadline;
+import Duke.Task.Event;
+import Duke.Task.Task;
+
+public class DateOfCommand extends Command {
+
+    private String description;
+    private TaskList taskList;
+    public DateOfCommand(String description, TaskList taskList) {
```
Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java:14]: Missing bank line separation?

Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java:14]: Having a bank line separation helps in enhancing readability

File: src/main/java/Duke/Command/DateOfCommand.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public String reply() {
+        if (description.length() == 0) {
+            throw new DukeIncompleteException();
+        }
+        try {
+            int index = Integer.parseInt(description);
+            Task task = taskList.get(index);
+            String type = (task instanceof Deadline) ? "Deadline" :
+                    (task instanceof Event) ? "Event" : "Todo";
+            if (type.equals("Todo")) {
```
Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java]: Having a bank line separation helps in enhancing readability

File: src/main/java/Duke/Command/DateOfCommand.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                return task.getDate() + index;
+            } else {
+                return "The " + type + " time of task " +
```
Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java:33]: The "+" sign should be at the next line as per coding standard

Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java]: The "+" sign at the end should be at the next line as per coding standard

File: src/main/java/Duke/Command/DateOfCommand.java (hunk 1 above)
Comment (dannytayjy) [reply src/main/java/Duke/Command/DateOfCommand.java:14]: I noticed the same issue in several other places too

Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java]: Having a bank line separation between the variable and the constructor helps in enhancing readability. I noticed the same issue in several other places too.

File: src/main/java/Duke/Command/DateOfCommand.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        this.description = description;
+        this.taskList = taskList;
+    }
+
+    @Override
+    public String reply() {
+        if (description.length() == 0) {
+            throw new DukeIncompleteException();
+        }
+        try {
```
Comment (dannytayjy) [inline src/main/java/Duke/Command/DateOfCommand.java]: Logical units within a block should be separated by one blank line. I noticed the same issue in several other places too.

File: src/main/java/Duke/Command/DateOfCommand.java (hunk 3 above)
Comment (dannytayjy) [reply src/main/java/Duke/Command/DateOfCommand.java]: I noticed the same issue in several other places too

File: src/main/java/Duke/Command/FindCommand.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        StringBuilder output = new StringBuilder("The following task(s) matches query '" +
+                description + "':\n");
+        int count = 0;
+        for (Task currentTask : taskList) {
+            if (currentTask.getTaskName().contains(description)) {
+                output.append(count + 1).append(". ").append(currentTask).append("\n");
+                count++;
+            }
+        }
+        if (count == 0) {
```
Comment (dannytayjy) [inline src/main/java/Duke/Command/FindCommand.java]: Having bank line separation between different blocks helps in enhancing readability

**Review by dannytayjy — COMMENTED** [review]
Your code quality seems okay to me but I may have missed out on certain things as it was getting hard to read without blank line separation between different blocks.

It would be good to have blank line separation as it helps in readability.

File: src/main/java/Duke/DukeException/DukeException.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public enum Type {INCOMPLETE, OUT_OF_BOUND, SYNTAX_ERROR, INVALID_FORMAT, WRONG_COMMAND}
+
+    private final String DEFAULT_STRING = "Please type in 'help' for instructions";
+
+    protected final Type type;
+
+    /**
+     * Constructor for Duke Exception class
+     * @param message Error message
```
Comment (YaleChen299) [inline src/main/java/Duke/DukeException/DukeException.java]: Maybe adding a line after the brief description before the parameter section would help to standardize the java doc. This applies to other places too.

File: src/main/java/Duke/Main/Ui.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package Duke.Main;
+
+import java.util.Scanner;
+
+public class Ui {
+
+    private final Scanner sc;
+
+    /**
+     * Constructor for the UI class, setting up a Scanner object
```
Comment (YaleChen299) [inline src/main/java/Duke/Main/Ui.java]: According to the standard in this course, it is better to add a full stop at the end of the brief description.

**Review by YaleChen299 — COMMENTED** [review]
Generally having good coding quality. However, maybe you can improve by adding more java doc and standardizing the them with the course coding standard. 😉

File: src/main/java/Duke/Command/DateOfCommand.java (hunk 2 above)
Comment (hmanhduc2k) [reply src/main/java/Duke/Command/DateOfCommand.java]: Thanks for the suggestion, I will consider adding a blank line :)

File: src/main/java/Duke/Command/DateOfCommand.java (hunk 3 above)
Comment (hmanhduc2k) [reply src/main/java/Duke/Command/DateOfCommand.java]: Noted, with thanks

File: src/main/java/Duke/DukeException/DukeException.java (hunk 6 above)
Comment (hmanhduc2k) [reply src/main/java/Duke/DukeException/DukeException.java]: Thanks for your suggestion, I will try to change how my JavaDoc is like
